File: 2021/21/python/solution.py
#!/usr/bin/env pypy3

from collections import Counter
from functools import lru_cache
from itertools import cycle
from itertools import product

DICE = sorted(Counter(map(sum, product(range(1, 4), repeat=3))).items())


class Player:
    def __init__(self, position, name):
        self.name = name
        self.position = position - 1
        self.score = 0

    def roll(self, roll):
        self.position = (self.position + roll) % 10
        self.score += self.position + 1


def read_input(filename):
    with open(filename) as lines:
        return [Player(int(line.split()[-1]), i) for i, line in enumerate(lines)]


def chunked(xs, n):
    xs = iter(xs)
    return zip(*[xs] * n)


def part_1(players):
    die = cycle(range(1, 101))
    rolls = map(sum, chunked(die, 3))
    for i, (player, roll) in enumerate(zip(cycle(players), rolls), 1):
        player.roll(roll)
        if player.score >= 1000:
            break

    return i * 3 * min(player.score for player in players)


@lru_cache(maxsize=None)
def part_2(p1, p2, s1, s2):
    wins1, wins2 = 0, 0
    for roll, multiplicity in DICE:
        position = (p1 + roll) % 10
        score = s1 + position + 1
        if score >= 21:
            wins1 += multiplicity
        else:
            w2, w1 = part_2(p2, position, s2, score)
            wins1 += w1 * multiplicity
            wins2 += w2 * multiplicity
    return wins1, wins2


def main():
    print(part_1(read_input("input")))

    p1, p2 = read_input("input")
    # clear cache to not influence benchmarking
    part_2.cache_clear()
    result = part_2(p1.position, p2.position, 0, 0)
    print(max(result))


if __name__ == "__main__":
    main()

File: 2021/21/cpp/solution.cxx
#include <cstdint>
#include <cstddef>

#include <algorithm>
#include <array>
#include <iostream>
#include <utility>

#include <range/v3/numeric/accumulate.hpp>
#include <range/v3/view/chunk.hpp>
#include <range/v3/view/cycle.hpp>
#include <range/v3/view/enumerate.hpp>
#include <range/v3/view/iota.hpp>
#include <range/v3/view/transform.hpp>


constexpr auto dice = std::to_array<std::pair<std::uint64_t, std::uint64_t>>({
    {3, 1},
    {4, 3},
    {5, 6},
    {6, 7},
    {7, 6},
    {8, 3},
    {9, 1},
});

struct Player {
    std::uint64_t name;
    std::uint64_t position;
    std::uint64_t score;

    auto roll(std::uint64_t const roll) -> void {
        this->position = (this->position + roll) % 10;
        this->score += this->position + 1;
    }
};

auto part_1(auto&& players) -> std::uint64_t {
    auto sides = ranges::views::iota(std::uint64_t{1}, std::uint64_t{101});
    auto die = sides | ranges::views::cycle;
    auto rolls = die
        | ranges::views::chunk(3)
        | ranges::views::transform(
            [](auto const& r) { return ranges::accumulate(r, std::uint64_t{0}); }
        );
    for (auto const& [i, roll]: rolls | ranges::views::enumerate) {
        auto& player = players[static_cast<std::size_t>(i) % 2];
        player.roll(roll);
        if (player.score >= 1000) {
            return (static_cast<std::uint64_t>(i) + 1) * 3 * players[(player.name + 1) % 2].score;
        }
    }
    std::unreachable();
}

auto part_2(
    std::uint64_t const p1, std::uint64_t const p2,
    std::uint64_t const s1, std::uint64_t const s2
) -> std::pair<std::uint64_t, std::uint64_t> {
    auto wins1 = std::uint64_t{0};
    auto wins2 = std::uint64_t{0};
    for (auto const& [roll, multiplicity]: dice) {
        auto const position = (p1 + roll) % 10;
        auto const score = s1 + position + 1;
        if (score >= 21) {
            wins1 += multiplicity;
        }
        else {
            auto const [w2, w1] = part_2(p2, position, s2, score);
            wins1 += w1 * multiplicity;
            wins2 += w2 * multiplicity;
        }
    }
    return std::pair{wins1, wins2};
}

auto main() -> int {
    std::cout << part_1(std::array{
        Player{.name = 0, .position = 2, .score = 0},
        Player{.name = 1, .position = 9, .score = 0},
    }) << "\n";
    auto solution = part_2(2, 9, 0, 0);
    std::cout << std::max(solution.first, solution.second) << "\n";
}
